### DISCUSSION TITLE
What is the best way to centrally manage default enabled features for all tenants after initialization.

### DISCUSSION THREAD

**Discussion by rikbosch** [OrchardCMS/OrchardCore#14835] [Q&A]
We're looking for a correct strategy to centrally manage enabled features for tenants, keeping in mind these will change in the future. The Default tenant has it's own feature set.

I know we can use setup recipe's to create the tenant and enable the features, but how would you enable a default feature after a tenant was initialized?

I can think of 2 options:
* Create 2 Default FeatureSet features, a `Default Tenant FeatureSet` and `Tenant FeatureSet` and use `DataMigration` to enable extra features. 
* Just enable the features using code :
```csharp
builder.Services
    .AddOrchardCms(orchardCoreBuilder =>
    {
           orchardCoreBuilder.AddTenantFeatures("Some.FeatureId");
    }
```

**Comment by MikeAlhayek** [comment]
OC already has a way to use the default tenant to manage the feature for other tenants. Something was added 8 months ago or so.

From the default tenant, list your tenants. Next to each tenant there is "Features" button that you can click and manage the features of that tenant. You can even disable the "Features" module for your tenant so you are the only one able to change their features.

**Reply by hishamco** [reply]
Also, you can enable/disable the tenants' features programmatically

**Reply by rikbosch** [reply]
Yes, but this still requires someone to manually disable/ enable/ reconfigure the features. We're managing a lot of (similar) tenants and want to reduce the manual configuration by automating (and testing) this kind of maintenance actions.

We're looking for a way to be able to automate and test feature upgrades for tenants.

But it sounds like a DataMigration might be the way to do this, as this runs in a Tenant's context and allows for custom code to execute.

**Reply by hishamco** [reply]
> Yes, but this still requires someone to manually disable/ enable/ reconfigure the features.

Why? you can enable / disable from within APIs

**Reply by weirdyang** [reply]
How do you enable/disable features programmatically?

**Reply by hishamco** [reply]
```csharp
ShellFeaturesManager.EnableFeaturesAsync("featuresToEnable");
```

**Reply by weirdyang** [reply]
thanks!

**Reply by weirdyang** [reply]
does this still work? the method seems to be asking for FeatureInfo class, and doesn't accept a string

![image](https://github.com/OrchardCMS/OrchardCore/assets/25576022/f06e1a19-6e15-47c8-9175-aacbff476b45)

**Reply by MikeAlhayek** [reply]
@weirdyang use the `var feature = await featureService.GetAvailableFeature(id);` method to get IFeatureInfo then use that to enable/disable features

**Comment by hishamco** [comment]
@rikbosch is this has been answered?

**Reply by rikbosch** [reply]
Not yet, but when I continue on this I'll add the strategy we picked here as well.
